### PR TITLE
Adding experimentaldata views

### PR DESCRIPTION
**Pull request by skeating**
The first part of #131.

### CODE REVIEW

**Comment by jonc125** [comment]
I tend only to ask for reviews when it's at a stage where you want feedback, which doesn't appear to be the case for this yet!

**Comment by skeating** [comment]
> I tend only to ask for reviews when it's at a stage where you want feedback, which doesn't appear to be the case for this yet!

But I didn't press the request button - I only listed potential reviewers ;-)

**Comment by jonc125** [comment]
> > I tend only to ask for reviews when it's at a stage where you want feedback, which doesn't appear to be the case for this yet!
> 
> But I didn't press the request button - I only listed potential reviewers ;-)

Hmm, the effect though was that I got an email asking me to review the PR!

File: weblab/datasets/models.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    # So instead we store the full git SHA as a string.
+    model_version = models.CharField(max_length=50)
+    protocol_version = models.CharField(max_length=50)
+
+    class Meta:
+        unique_together = ('model', 'protocol', 'model_version', 'protocol_version')
+        verbose_name_plural = 'ExperimentalDatasets'
+
+        permissions = (
+            ('create_dataset_experiment', 'Can create dataset experiments'),
```
[the remark this answers is not in the document]
Comment (jonc125) [reply weblab/datasets/models.py]: Or even just `('create_dataset', 'Can create experimental datasets')` to match with the template tag below.

File: weblab/datasets/urls.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from django.conf.urls import url
+
+from . import views
+
+
+_COMMIT = r'(?P<sha>[^^~:/ ]+)'
+_FILENAME = r'(?P<filename>[\w\-. \%:]+)'
+_FILEVIEW = r'%s/(?P<viz>\w+)' % _FILENAME
+
+urlpatterns = [
```
Comment (jonc125) [inline weblab/datasets/urls.py]: These look good, and yes, we'll need the commented ones too in due course.

File: weblab/datasets/urls.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+from django.conf.urls import url
+
+from . import views
+
+
+_COMMIT = r'(?P<sha>[^^~:/ ]+)'
```
Comment (jonc125) [inline weblab/datasets/urls.py]: We shouldn't need the `_COMMIT` but the others will probably form part of the detail view url.

File: weblab/datasets/views.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    UserFormKwargsMixin, CreateView
+):
+    """
+    Create new ExperimentalDataset
+    """
+    template_name = 'datasets/dataset_form.html'
+
+    @property
+    def permission_required(self):
+        datasets.create_dataset_experiment
```
Comment (jonc125) [inline weblab/datasets/views.py]: Does this need a `return`?

Comment (jonc125) [reply weblab/datasets/views.py]: Or just `permission_required = 'datasets.create_dataset'` without the method?

File: weblab/datasets/views.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    def get_queryset(self):
+        return self.model.objects.filter(author=self.request.user)
+
+
+# class ExperimentalDatasetView(VisibilityMixin, SingleObjectMixin, RedirectView):
+#     """
+#     View an ExperimentalDataset
+#
+#     All this does is redirect to the latest version of the entity, if it exists.
```
Comment (jonc125) [inline weblab/datasets/views.py]: We don't have versions for datasets, so this should *be* the equivalent of the latest version view.

File: weblab/datasets/views.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+from .forms import (
+    FileUploadForm,
+)
+
+class ExperimentalDatasetCreateView(
+    LoginRequiredMixin, PermissionRequiredMixin,
+    UserFormKwargsMixin, CreateView
+):
+    """
+    Create new ExperimentalDataset
```
Comment (jonc125) [inline weblab/datasets/views.py]: Given we don't have versions, this is probably going to be more complex!

File: weblab/templates/datasets/dataset_form.html [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{% extends "base.html" %}
```
Comment (jonc125) [inline weblab/templates/datasets/dataset_form.html]: Should this file be called `dataset_create.html`?

File: weblab/templates/datasets/dataset_form.html [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+{% block content %}
+<div id='newdatasetform'>
+  <h1>Create {{ type }}</h1>
+
+  <form method="POST" action="">
+    {% csrf_token %}
+
+    {{ form.as_p }}
+
```
Comment (jonc125) [inline weblab/templates/datasets/dataset_form.html]: Some of the 'new version' stuff for entities will need to come in here too.

File: weblab/templates/datasets/dataset_list.html [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+{% load datasets %}
+
+{% block title %}Your datasets - {% endblock title %}
+
+{% block content %}
+
+
+    {% can_create_dataset as permission %}
+    {% if permission %}
+    <small><a href="{{ type|url_new }}" id="add{{ type }}link" class="pointer">Create a new dataset</a></small>
```
Comment (jonc125) [inline weblab/templates/datasets/dataset_list.html]: Probably don't need the `type` in the file as its always used for experimental datasets.

File: weblab/templates/datasets/dataset_list.html [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    {% else %}
+    <small>Your account doesn't have the authority to upload {{ type }}s; please <a href="{% url 'contact' %}">contact us</a> to request permission.</small>
+    {% endif %}
+
+    <ul>
+      {% for entity in object_list %}
+        <li title="{{ entity.name }}">
+          <strong><a href="{{ entity|url_entity }}">{{ entity.name }}</a></strong>
+
+          {% can_create_version entity as permission %}
```
Comment (jonc125) [inline weblab/templates/datasets/dataset_list.html]: This bit will go.

File: weblab/templates/datasets/dataset_list.html [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    {% if permission %}
+    <small><a href="{{ type|url_new }}" id="add{{ type }}link" class="pointer">Create a new dataset</a></small>
+    {% else %}
+    <small>Your account doesn't have the authority to upload {{ type }}s; please <a href="{% url 'contact' %}">contact us</a> to request permission.</small>
+    {% endif %}
+
+    <ul>
+      {% for entity in object_list %}
+        <li title="{{ entity.name }}">
+          <strong><a href="{{ entity|url_entity }}">{{ entity.name }}</a></strong>
```
Comment (jonc125) [inline weblab/templates/datasets/dataset_list.html]: Need to change the url :)

File: weblab/datasets/apps.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
+from django.apps import AppConfig
+from django.db.models.signals import post_save, pre_delete
+
+#from .signals import entity_created, entity_deleted
+
+
+class DatasetsConfig(AppConfig):
+    name = 'datasets'
+
+    def ready(self):
```
[the remark this answers is not in the document]
Comment (skeating) [reply weblab/datasets/apps.py]: left file but no code in

File: weblab/datasets/models.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+from core.models import VisibilityModelMixin
+from entities.models import ModelEntity, ProtocolEntity
+
+class ExperimentalDataset(UserCreatedModelMixin, VisibilityModelMixin, models.Model):
+    """Prototyping class for experimental datasets
+    """
+    model = models.ForeignKey(ModelEntity, related_name='model_experimental_dataset')
+    protocol = models.ForeignKey(ProtocolEntity, related_name='protocol_experimental_dataset')
+
+    # Note that we can't use a ForeignKey here, because versions of models and protocols
```
[the remark this answers is not in the document]
Comment (skeating) [reply weblab/datasets/models.py]: I assume this means the versions

File: weblab/datasets/signals.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
+
```
[the remark this answers is not in the document]
Comment (skeating) [reply weblab/datasets/signals.py]: but are we not going to create/delete datasets in the same way as entities - ie file upload ??

File: weblab/templates/datasets/dataset_form.html (hunk 7 above)
Comment (skeating) [reply weblab/templates/datasets/dataset_form.html]: copied from entity_form so I don't know :-)

File: weblab/templates/datasets/dataset_list.html (hunk 10 above)
Comment (skeating) [reply weblab/templates/datasets/dataset_list.html]: Presumably this should list datasets here

Comment (skeating) [reply weblab/templates/datasets/dataset_list.html]: Okay I see you mean the permission to create aversion bit will go

File: weblab/templates/datasets/dataset_form.html (hunk 7 above)
Comment (jonc125) [reply weblab/templates/datasets/dataset_form.html]: Indeed, not sure why it's called that there either :) Especially as I think there's a line in the view specifying the template name anyway!

File: weblab/datasets/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jonc125) [reply weblab/datasets/models.py]: Yes

File: weblab/datasets/models.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+class ExperimentalDataset(UserCreatedModelMixin, VisibilityModelMixin, models.Model):
+    """Prototyping class for experimental datasets
+    """
+    protocol = models.ForeignKey(ProtocolEntity, related_name='protocol_experimental_datasets')
+
+    class Meta:
+        verbose_name_plural = 'ExperimentalDatasets'
+
+        permissions = (
+            ('create_dataset', 'Can create experiment datasets'),
```
Comment (jonc125) [inline weblab/datasets/models.py]: experimental not experiment

File: weblab/datasets/signals.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jonc125) [reply weblab/datasets/signals.py]: Yes, but we don't need to initialise a git repo, so shouldn't need the create bit. We probably do need the delete handing though!

File: weblab/datasets/views.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+from repocache.exceptions import RepoCacheMiss
+from repocache.models import CachedEntityVersion
+
+from .models import ExperimentalDataset
+
+from .forms import (
+    FileUploadForm,
+)
+
+class ExperimentalDatasetMixin:
```
Comment (jonc125) [inline weblab/datasets/views.py]: I don't think this is needed. Instead each view should have a class attribute
```python
    model = ExperimentalDataset
```

File: weblab/datasets/views.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    @property
+    def form_class(self):
+        return FormMixin
+
+    def get_success_url(self):
+        return reverse('datasets:newversion',
+                       args=[self.object.pk])
+
+
+class ExperimentalDatasetListView(LoginRequiredMixin, ListView, ExperimentalDatasetMixin):
```
Comment (jonc125) [inline weblab/datasets/views.py]: As mentioned, you shouldn't need `ExperimentalDatasetMixin`. Also I think `ListView` needs to come last.

File: weblab/templates/datasets/dataset_list.html [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+{% load datasets %}
+
+{% block title %}Your datasets - {% endblock title %}
+
+{% block content %}
+
+
+    {% can_create_dataset as permission %}
+    {% if permission %}
+    <small><a href="{{ url_new }}" id="add dataset link" class="pointer">Create a new dataset</a></small>
```
Comment (jonc125) [inline weblab/templates/datasets/dataset_list.html]: You can probably use `{% url 'datasets:new' %}` in place of `{{url_new}}`

File: weblab/templates/includes/navbar.html [hunk 19]
```diff
@@ -6,6 +6,7 @@
       <li><a href="{% url 'experiments:list' 'mine' %}">Experiments</a></li>
       <li><a href="{% url 'entities:list' 'model' %}">My models</a></li>
       <li><a href="{% url 'entities:list' 'protocol' %}">My protocols</a></li>
+      <li><a href="{% url 'datasets:list' 'mine' %}">My datasets</a></li>
```
Comment (jonc125) [inline weblab/templates/includes/navbar.html]: I don't think the `'mine'` is wanted here?

Comment (jonc125) [reply weblab/templates/includes/navbar.html]: Ah, no, I see, you're thinking to allow listing not just their own datasets in due course. OK, makes sense.

**Pull request merged**